### PR TITLE
Add auditing for all user actions #3078

### PR DESCRIPTION
**Pull request by maartenba**
Thuis PR ensures the following operations are audited:

* User audit actions:
  * Registered
  * AddedCredential
  * RemovedCredential
  * RequestedPasswordReset
  * ChangeEmail
  * CancelChangeEmail
  * ConfirmEmail
  * Login
  * Failed logins
* Package audit actions:
  * Deleted
  * SoftDeleted
  * Publish
  * List/Unlist
  * Edit metadata
  * Failed push due to invalid credential
* Package registration audit actions:
  * Add/remove owner

Please have a look @skofman1 @xavierdecoster

### CODE REVIEW

File: src/NuGetGallery.Core/Auditing/AuditActor.cs [hunk 1]
```diff
@@ -26,6 +26,7 @@ public AuditActor(string machineName, string machineIP, string userName, string
         public AuditActor(string machineName, string machineIP, string userName, string authenticationType, DateTime timeStampUtc, AuditActor onBehalfOf)
         {
             MachineName = machineName;
+            MachineIP = machineIP; // TODO: can we store this information?
```
Comment (xavierdecoster) [inline src/NuGetGallery.Core/Auditing/AuditActor.cs]: None issue for server IPs. If client IP, this is PII, which means "to investigate" :)

Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/AuditActor.cs]: @yishaigalatzer do you know if we can store client IP for auditing purposes?

File: src/NuGetGallery/Controllers/ApiController.cs [hunk 2]
```diff
@@ -38,10 +39,12 @@ public partial class ApiController
...
         public IMessageService MessageService { get; set; }
+        public AuditingService AuditingService { get; set; }
         public ConfigurationService ConfigurationService{ get; set; }
 
         protected ApiController()
         {
+            AuditingService = AuditingService.None;
         }
 
         public ApiController(
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/ApiController.cs]: do we need so many ctor for one class?

File: src/NuGetGallery/Controllers/ApiController.cs [hunk 3]
```diff
@@ -317,6 +323,10 @@ public virtual Task<ActionResult> CreatePackagePost()
                             IndexingService.UpdatePackage(package);
                         }
 
+                        // Write an audit record
+                        await AuditingService.SaveAuditRecord(new PackageAuditRecord(package, PackageAuditAction.Created, "Package created from API."));
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/ApiController.cs]: can we make this a const, or even an enum of "package creator" (or better name)

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 4]
```diff
@@ -1165,6 +1174,10 @@ private ActionResult EditFailed(string id, EditPackageRequest formData, Package
                 // tell Lucene to update index for the new package
                 _indexingService.UpdateIndex();
 
+                // write an audit record
+                await _auditingService.SaveAuditRecord(new PackageAuditRecord(package, PackageAuditAction.Created, "Package created from web interface."));
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/PackagesController.cs]: please use const or enum

File: src/NuGetGallery/Services/PackageService.cs [hunk 5]
```diff
@@ -20,19 +21,22 @@ public class PackageService : IPackageService
...
         public PackageService(
             IEntityRepository<PackageRegistration> packageRegistrationRepository,
             IEntityRepository<Package> packageRepository,
             IEntityRepository<PackageOwnerRequest> packageOwnerRequestRepository,
             IIndexingService indexingService,
-            IPackageNamingConflictValidator packageNamingConflictValidator)
+            IPackageNamingConflictValidator packageNamingConflictValidator,
+            AuditingService auditingService)
         {
             _packageRegistrationRepository = packageRegistrationRepository;
```
Comment (skofman1) [inline src/NuGetGallery/Services/PackageService.cs]: null checks

File: src/NuGetGallery/Services/PackageService.cs [hunk 6]
```diff
@@ -244,6 +248,9 @@ where VersionRange.Parse(d.VersionSpec).Satisfies(packageVersion)
                 _packageOwnerRequestRepository.DeleteOnCommit(request);
                 await _packageOwnerRequestRepository.CommitChangesAsync();
             }
+            
+            await _auditingService.SaveAuditRecord(
+                new PackageRegistrationAuditRecord(package, PackageRegistrationAuditAction.AddOwner, $"Added owner: {user.Username}"));
```
Comment (skofman1) [inline src/NuGetGallery/Services/PackageService.cs]: why do we need the message $"Added Owner", isn't the context clear from the enum PackageRegistrationAuditAction.AddOwner?

File: src/NuGetGallery/Services/PackageService.cs [hunk 7]
```diff
@@ -263,6 +270,9 @@ where VersionRange.Parse(d.VersionSpec).Satisfies(packageVersion)
 
             package.Owners.Remove(user);
             await _packageRepository.CommitChangesAsync();
+
+            await _auditingService.SaveAuditRecord(
+                new PackageRegistrationAuditRecord(package, PackageRegistrationAuditAction.RemoveOwner, $"Removed owner: {user.Username}"));
```
Comment (skofman1) [inline src/NuGetGallery/Services/PackageService.cs]: same

File: src/NuGetGallery/Controllers/ApiController.cs [hunk 8]
```diff
@@ -38,10 +39,12 @@ public partial class ApiController
         public IAutomaticallyCuratePackageCommand AutoCuratePackage { get; set; }
         public IStatusService StatusService { get; set; }
         public IMessageService MessageService { get; set; }
+        public AuditingService AuditingService { get; set; }
         public ConfigurationService ConfigurationService{ get; set; }
 
         protected ApiController()
         {
+            AuditingService = AuditingService.None;
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/ApiController.cs]: you could use a factory instead...

File: src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    public class FileSystemAuditingService : AuditingService
+    {
+        public static readonly string DefaultContainerName = "auditing";
+
+        private readonly string _auditingPath;
+        private readonly Func<Task<AuditActor>> _onBehalfOfThunk;
+        
+        public FileSystemAuditingService(string auditingPath, Func<Task<AuditActor>> onBehalfOfThunk)
+        {
+            _auditingPath = auditingPath;
```
Comment (skofman1) [inline src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs]: null checks

**Comment by skofman1** [comment]
>             // Track the value for credentials that are definitely revokable (API Key, etc.) and have been removed

I'm confused.. do we save the Api Key value? Why? Are we allowed?

---
Refers to: src/NuGetGallery.Core/Auditing/CredentialAuditRecord.cs:19 in d1d291b. [](commit_id = d1d291bd9b501df7be94db1e84f953558f157109, deletion_comment = False)

**Comment by yishaigalatzer** [comment]
Also add failed login attempts (if not there) and failed publish (because of ApiKey)

File: src/NuGetGallery.Core/Auditing/AuditActor.cs (hunk 1 above)
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/AuditActor.cs]: Lets follow up offline with privacy/security folk

File: src/NuGetGallery.Core/Auditing/AuditingService.cs [hunk 10]
```diff
@@ -71,8 +68,8 @@ private class NullAuditingService : AuditingService
         {
             protected override Task<Uri> SaveAuditRecord(string auditData, string resourceType, string filePath, string action, DateTime timestamp)
             {
-                var uriString = $"http://auditing.local/{resourceType}/{filePath}/{timestamp:s}-{action.ToLowerInvariant()}";
-                var uri = new Uri(uriString);
+                var uri = new Uri($"http://auditing.local/{resourceType}/{filePath}/{timestamp:s}-{action.ToLowerInvariant()}");
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/AuditingService.cs]: why the change? also can you break line 70 so it fits on the screen without scrolling?

File: src/NuGetGallery.Core/Auditing/CredentialAuditRecord.cs [hunk 11]
```diff
@@ -5,8 +5,14 @@ namespace NuGetGallery.Auditing
 {
     public class CredentialAuditRecord
     {
+        public int Key { get; set; }
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/CredentialAuditRecord.cs]: Do these need to be mutable?

File: src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+                return null;
+            }
+
+            // Try to identify the client IP using various server variables
+            var clientIpAddress = context.Request.ServerVariables["HTTP_X_FORWARDED_FOR"];
+            if (string.IsNullOrEmpty(clientIpAddress)) // Try REMOTE_ADDR server variable
+            {
+                clientIpAddress = context.Request.ServerVariables["REMOTE_ADDR"];
+            }
+            if (string.IsNullOrEmpty(clientIpAddress)) // Try UserHostAddress property
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs]: newline after `}`

File: src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        private readonly string _auditingPath;
+        private readonly Func<Task<AuditActor>> _onBehalfOfThunk;
+        
+        public FileSystemAuditingService(string auditingPath, Func<Task<AuditActor>> onBehalfOfThunk)
+        {
+            _auditingPath = auditingPath;
+            _onBehalfOfThunk = onBehalfOfThunk;
+        }
+
+        public static Task<AuditActor> AspNetActorThunk()
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs]: this is not a thunk :)

File: src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+                action.ToLowerInvariant(), ".audit.v1.json");
+
+            // Build full file path
+            var fullFilePath = Path.Combine(_auditingPath, relativeFilePath);
+
+            // Ensure the directory exists
+            var directoryName = Path.GetDirectoryName(fullFilePath);
+            if (!Directory.Exists(directoryName))
+            {
+                // TODO: should we catch the exception and log? or throw like we do now?
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs]: No TODO: in code

File: src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+                // in my opinion, auditing is super important and should throw on failure
+                Directory.CreateDirectory(directoryName);
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs]: If we throw, would that fail the whole operation?

File: src/NuGetGallery.Core/Auditing/PackageAuditAction.cs [hunk 16]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGetGallery.Auditing
+{
+    public enum PackageAuditAction
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/PackageAuditAction.cs]: Can we call it just PackageAction or AuditedPackageAction

File: src/NuGetGallery.Core/Auditing/PackageAuditAction.cs [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    {
+        Deleted,
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/PackageAuditAction.cs]: Delete, SoftDelete, Create, ...

File: src/NuGetGallery.Core/Auditing/PackageAuditRecord.cs [hunk 18]
```diff
@@ -28,21 +27,24 @@ public PackageAuditRecord(string id, string version, string hash, DataTable pack
             Reason = reason;
         }
 
-        public PackageAuditRecord(Package package, DataTable packageRecord, DataTable registrationRecord, PackageAuditAction action, string reason)
-            : this(package.PackageRegistration.Id, package.Version, package.Hash, packageRecord, registrationRecord, action, reason)
+        public PackageAuditRecord(Package package, PackageAuditAction action, string reason)
+            : this(package.PackageRegistration.Id, package.Version, package.Hash, null, null, action, reason)
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/PackageAuditRecord.cs]: can we use named parameters for the nulls?

File: src/NuGetGallery.Core/Auditing/PackageRegistrationAuditAction.cs [hunk 19]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGetGallery.Auditing
+{
+    public enum PackageRegistrationAuditAction
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/PackageRegistrationAuditAction.cs]: same comment for naming as above.

File: src/NuGetGallery.Core/Auditing/PackageRegistrationAuditRecord.cs [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+namespace NuGetGallery.Auditing
+{
+    public class PackageRegistrationAuditRecord : AuditRecord<PackageRegistrationAuditAction>
+    {
+        public string Id { get; set; }
+        public DataTable RegistrationRecord { get; set; }
+
+        public string Reason { get; set; }
+
+        public PackageRegistrationAuditRecord(string id,  DataTable registrationRecord, PackageRegistrationAuditAction action, string reason)
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/PackageRegistrationAuditRecord.cs]: break long lines

File: src/NuGetGallery.Core/Auditing/PackageRegistrationAuditRecord.cs [hunk 21]
```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Data;
+
+namespace NuGetGallery.Auditing
+{
+    public class PackageRegistrationAuditRecord : AuditRecord<PackageRegistrationAuditAction>
+    {
+        public string Id { get; set; }
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/PackageRegistrationAuditRecord.cs]: get only

File: src/NuGetGallery.Core/Auditing/UserAuditAction.cs [hunk 22]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGetGallery.Auditing
+{
+    public enum UserAuditAction
+    {
+        Registered,
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/Auditing/UserAuditAction.cs]: Register, AddCredential, ...

File: src/NuGetGallery.Core/DataTableUtility.cs [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+using System.Data;
+using System.Globalization;
+
+namespace NuGetGallery
+{
+    public static class DataTableUtility
+    {
+        public static DataTable ConvertToDataTable<T>(T instance)
+        {
+            PropertyDescriptorCollection properties = TypeDescriptor.GetProperties(typeof(T));
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/DataTableUtility.cs]: This could potentially be super expensive if you don't cache. Can you please profile?

File: src/NuGetGallery.Core/DataTableUtility.cs [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+            }
+
+            table.Rows.Add(values.ToArray());
+
+            return table;
+        }
+
+        private static bool IsComplexType(Type type)
+        {
+            if (type.IsSubclassOf(typeof(ValueType)) || type == typeof(string))
```
Comment (yishaigalatzer) [inline src/NuGetGallery.Core/DataTableUtility.cs]: `return ! type.IsSubclassOf(typeof(ValueType)) || type == typeof(string))`

File: src/NuGetGallery/Controllers/ApiController.cs (hunk 3 above)
Comment (yishaigalatzer) [inline src/NuGetGallery/Controllers/ApiController.cs]: and break the super long lines

File: src/NuGetGallery/RequestModels/EditPackageVersionRequest.cs [hunk 25]
```diff
@@ -103,5 +108,22 @@ public EditPackageVersionRequest(Package package, PackageEdit pendingMetadata)
 
         [Display(Name = RequiresLicenseAcceptanceStr)]
         public bool RequiresLicenseAcceptance { get; set; }
+
+        public Package Simulate(Package package)
```
Comment (yishaigalatzer) [inline src/NuGetGallery/RequestModels/EditPackageVersionRequest.cs]: Please add comments to this method. I don't understand what Simulate means from reading the code

File: src/NuGetGallery/Services/PackageService.cs (hunk 6 above)
Comment (yishaigalatzer) [inline src/NuGetGallery/Services/PackageService.cs]: I agree with Svetlana, if the messages are more specific, it will also be easier to get data out of them

**Comment by maartenba** [comment]
@skofman RE:https://github.com/NuGet/NuGetGallery/pull/3083#issuecomment-226820714 - the value is only stored when: 1) it is an API key and 2) it has just been replaced. Audits will never contain API keys that are in use. That said, this used to be the behavior and we're free to update if this makes no sense.

File: src/NuGetGallery.Core/Auditing/AuditActor.cs (hunk 1 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/AuditActor.cs]: Seems we can store it if we obfuscate the last octet. Updated the code to reflect this. (replace last octet with 0)

File: src/NuGetGallery/Controllers/ApiController.cs (hunk 2 above)
Comment (maartenba) [inline src/NuGetGallery/Controllers/ApiController.cs]: Yeah the tests use these in various locations.

File: src/NuGetGallery/Controllers/ApiController.cs (hunk 3 above)
Comment (maartenba) [inline src/NuGetGallery/Controllers/ApiController.cs]: Fixed

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 4 above)
Comment (maartenba) [inline src/NuGetGallery/Controllers/PackagesController.cs]: Fixed

File: src/NuGetGallery/Services/PackageService.cs (hunk 5 above)
Comment (maartenba) [inline src/NuGetGallery/Services/PackageService.cs]: Fixed

File: src/NuGetGallery/Services/PackageService.cs (hunk 6 above)
Comment (maartenba) [inline src/NuGetGallery/Services/PackageService.cs]: Switched to passing in the owner name to the audit record directly.

File: src/NuGetGallery/Services/PackageService.cs (hunk 7 above)
Comment (maartenba) [inline src/NuGetGallery/Services/PackageService.cs]: Fixed

File: src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs (hunk 9 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs]: Fixed

File: src/NuGetGallery.Core/Auditing/AuditingService.cs (hunk 10 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/AuditingService.cs]: Fixed

File: src/NuGetGallery.Core/Auditing/CredentialAuditRecord.cs (hunk 11 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/CredentialAuditRecord.cs]: Fixed

File: src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs (hunk 12 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs]: Fixed

File: src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs (hunk 14 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs]: Fixed

File: src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs (hunk 15 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/FileSystemAuditingService.cs]: Just the audit would fail. Not sure whether we should fail or not.

File: src/NuGetGallery.Core/Auditing/PackageAuditAction.cs (hunk 16 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/PackageAuditAction.cs]: Renamed

File: src/NuGetGallery.Core/Auditing/PackageAuditAction.cs (hunk 17 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/PackageAuditAction.cs]: Renamed. Note that existing audit records will stil have "Deleted" in the action name.

File: src/NuGetGallery.Core/Auditing/PackageAuditRecord.cs (hunk 18 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/PackageAuditRecord.cs]: Updated

File: src/NuGetGallery.Core/Auditing/PackageRegistrationAuditAction.cs (hunk 19 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/PackageRegistrationAuditAction.cs]: Updated

File: src/NuGetGallery.Core/Auditing/PackageRegistrationAuditRecord.cs (hunk 20 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/PackageRegistrationAuditRecord.cs]: Fixed

File: src/NuGetGallery.Core/Auditing/PackageRegistrationAuditRecord.cs (hunk 21 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/PackageRegistrationAuditRecord.cs]: Fixed

File: src/NuGetGallery.Core/Auditing/UserAuditAction.cs (hunk 22 above)
Comment (maartenba) [inline src/NuGetGallery.Core/Auditing/UserAuditAction.cs]: Fixed

File: src/NuGetGallery.Core/DataTableUtility.cs (hunk 23 above)
Comment (maartenba) [inline src/NuGetGallery.Core/DataTableUtility.cs]: Factored this out in favor of a simple DTO class.

File: src/NuGetGallery/RequestModels/EditPackageVersionRequest.cs (hunk 25 above)
Comment (maartenba) [inline src/NuGetGallery/RequestModels/EditPackageVersionRequest.cs]: Renamed + commented.

**Comment by maartenba** [comment]
@yishaigalatzer Added additional audits re: https://github.com/NuGet/NuGetGallery/pull/3083/#issuecomment-226903803

**Comment by maartenba** [comment]
Think we are good to merge. Can you have another pass?
